airflow/models/taskinstance.py

```diff
@@ -939,9 +939,9 @@ def next_retry_datetime(self):
         delay = self.task.retry_delay
         if self.task.retry_exponential_backoff:
             # If the min_backoff calculation is below 1, it will be converted to 0 via int. Thus,
-            # we must round up prior to converting to an int, otherwise a divide by zero error
-            # will occur in the modded_hash calculation.
-            min_backoff = int(math.ceil(delay.total_seconds() * (2 ** (self.try_number - 2))))
+            # we must impose a lower bound of 1 prior to converting to an int, otherwise a divide
+            # by zero error will occur in the modded_hash calculation.
+            min_backoff = int(max(1, delay.total_seconds() * (2 ** (self.try_number - 2))))
```

Comment:
This changes the value for (say) `delay=timedelta(seconds=1.5), self.try_number=2`. Previously the resul would be 2 (`ceil(1.5 * 1)`) but now it’s 1 (`int(1.5 * 1)`).

Reply:
Yeah, actually previously it was defined without the ceiling - the ceiling was added to deal with divide-by-zero issue. So I guess it is fine to revert to the previous definition, and just add a lower bound.

https://github.com/apache/airflow/commit/d1dceff69ab15f9ce28ea04215a693b81c8a93a2

Reply:
That change was made between 1.x and 2.x though and could be categorised as a breaking change. Not sure (one second isn’t really that important either way), maybe other maintainers can chime in on this.

Reply:
How about replacing the retry_delay value before the calculation?

```
modified_delay = delay.total_seconds() if delay.total_seconds() > 0 else 1
min_backoff = int(math.ceil(modified_delay * (2 ** (self.try_number - 2))))
```

example:
```
import math
import numpy as np
from datetime import timedelta

for i in np.arange(0, 3, 0.5):
    for try_number in range(1, 3, 1):
        delay = timedelta(seconds=i)

        min_backoff = int(math.ceil(delay.total_seconds() * (2 ** (try_number - 2))))

        modified_delay = delay.total_seconds() if delay.total_seconds() > 0 else 1
        min_backoff_new = int(math.ceil(modified_delay * (2 ** (try_number - 2))))

        print(f"Input(seconds: {i} , try_number: {try_number}) --> Result(Prev: {min_backoff}, New: {min_backoff_new})")

```
Output:
```
Input(seconds: 0.0 , try_number: 1) --> Result(Prev: 0, New: 1)
Input(seconds: 0.0 , try_number: 2) --> Result(Prev: 0, New: 1)
Input(seconds: 0.5 , try_number: 1) --> Result(Prev: 1, New: 1)
Input(seconds: 0.5 , try_number: 2) --> Result(Prev: 1, New: 1)
Input(seconds: 1.0 , try_number: 1) --> Result(Prev: 1, New: 1)
Input(seconds: 1.0 , try_number: 2) --> Result(Prev: 1, New: 1)
Input(seconds: 1.5 , try_number: 1) --> Result(Prev: 1, New: 1)
Input(seconds: 1.5 , try_number: 2) --> Result(Prev: 2, New: 2)
Input(seconds: 2.0 , try_number: 1) --> Result(Prev: 1, New: 1)
Input(seconds: 2.0 , try_number: 2) --> Result(Prev: 2, New: 2)
Input(seconds: 2.5 , try_number: 1) --> Result(Prev: 2, New: 2)
Input(seconds: 2.5 , try_number: 2) --> Result(Prev: 3, New: 3)
```

Reply:
That works. Is the reason for using `modified_delay` to prevent this from becoming a breaking change?

Reply:
What is the breaking change?
For `0` it never worked, for the rest of values they stay the same.

Reply:
I meant that my original solution is a breaking change, even though the difference is only one second. In practice, this is not a big difference, but I am curious how significant the breaking change is from a code perspective.

Reply:
Personally I don't think in this case 1 second is that important.
That said, we have another solution that avoid the need to debate/raise this question so why not take it? :)

Reply:
@edwardwang888 would you like to finish this PR?

Reply:
@eladkal Yes, sorry about the delay. I plan to finish it sometime this week - hopefully it should be a small change.

Reply:
@eladkal I have updated the PR. One test is failing but the failure looks unrelated to this change.

Reply:
I am good with the approach as is now. @eladkal ?

Reply:
Yep